Sources/SwiftLexicalLookup/ScopeSyntax.swift

```diff
@@ -0,0 +1,109 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+extension SyntaxProtocol {
+  /// Returns all names that `for` refers to at this syntax node.
+  /// Optional configuration can be passed as `with` to customize the lookup behavior.
+  ///
+  /// - Returns: An array of `LookupResult` for name `for` at this syntax node,
+  /// ordered by visibility. If set to `nil`, returns all available names ordered by visibility.
+  /// The order is from the innermost to the outermost scope,
+  /// and within each result, names are ordered by their introduction
+  /// in the source code.
+  ///
+  /// Example usage:
+  /// ```swift
+  /// class C {
+  ///   var a = 42
+  ///
+  ///   func a(a: Int) {
+  ///     a // <--- lookup here
+  ///
+  ///     let a = 0
+  ///   }
+  ///
+  ///   func a() {
+  ///     // ...
+  ///   }
+  /// }
+  /// ```
+  /// When calling this function on the declaration reference `a` within its name,
+  /// the function returns the parameter first, then the identifier of the variable
+  /// declaration, followed by the first function name, and then the second function name,
+  /// in this exact order. The constant declaration within the function body is omitted
+  /// due to the ordering rules that prioritize visibility within the function body.
+  @_spi(Experimental) public func lookup(
+    for name: String?,
```

Comment:
I think it could maybe be a bit tricky to also support implicit name lookup like `self`, `newValue` etc. when using `Identifier`. That’s why I actually had to remove `self` capture from this PR as implicit name handling is now part of #2748. I might’ve missed something, so I’d love to hear your take on this.

Reply:
I think `newValue` can be easily represented as an `Identifier`. We just need to add an initializer that creates a new `Identifier` from a `String`. 

I was briefly suspicious of `self` because I thought that the keyword `self` and the identifier `self` could refer to different things. But looks like they are considered identical

```swift
struct Foo {
  func test() {
    let `self` = 1
    print(self)
    print(`self`)
  }
}

Foo().test()
// Prints
// 1
// 1

// I might have expected
// Foo()
// 1
```

The following deviation of this might be a good test case as well

```swift
```

Reply:
I created a separate [branch with Identifier-based lookup](https://github.com/MAJKFL/swift-syntax/tree/identifier-based-lookup). I'm not entirely sure this is the most optimal solution, as it's a naive one, so I'd love to hear your feedback.

[I added a new `StringIdentifier` that strips backticks](https://github.com/MAJKFL/swift-syntax/blob/0ef6a21e325e7cd29b5f1d53be4953e4544c449c/Sources/SwiftSyntax/Identifier.swift#L71-L84) and added it as an optional property to `Identifier`. I also [made optional](https://github.com/MAJKFL/swift-syntax/blob/0ef6a21e325e7cd29b5f1d53be4953e4544c449c/Sources/SwiftSyntax/Identifier.swift#L21-L24) the original `raw` and `arena` properties. I'm mainly concerned about having the old properties as optional.

```swift
public struct StringIdentifier: Equatable, Hashable, Sendable {
  public let name: String

  fileprivate init(_ string: String) {
    let backtick = "`"
    if string.count > 2 && string.hasPrefix(backtick) && string.hasSuffix(backtick) {
      let startIndex = string.index(after: string.startIndex)
      let endIndex = string.index(before: string.endIndex)
      self.name = String(string[startIndex..<endIndex])
    } else {
      self.name = string
    }
  }
}
```

I also added the [test cases](https://github.com/MAJKFL/swift-syntax/blob/0ef6a21e325e7cd29b5f1d53be4953e4544c449c/Tests/SwiftLexicalLookupTest/NameLookupTests.swift#L703-L740) you've mentioned, that now work as expected due to the new Identifier-based name comparison.

As it's an internal change in `SwiftSyntax` module, I'd like to first ask for your feedback before adding it into the next PR. Can you perhaps think of a better way to handle this?